File: packages/puppeteer-parse/src/browser.ts
import { Browser, Target } from 'puppeteer-core'
import puppeteer from 'puppeteer-extra'
import AdblockerPlugin from 'puppeteer-extra-plugin-adblocker'
import StealthPlugin from 'puppeteer-extra-plugin-stealth'

if (process.env['USE_FIREFOX'] != 'true') {
  puppeteer.use(StealthPlugin())
  puppeteer.use(AdblockerPlugin({ blockTrackers: true }))
}

let browserInstance: Browser | null = null

const handleDisconnection = async () => {
  console.log('Browser disconnected, reconnecting...')
  browserInstance = null
  await getBrowser()
}

export const getBrowser = async (): Promise<Browser> => {
  if (browserInstance) {
    return browserInstance
  }

  console.log('Starting puppeteer browser')

  browserInstance = (await puppeteer.launch({
    args: [
      '--autoplay-policy=user-gesture-required',
      '--disable-component-update',
      '--disable-domain-reliability',
      '--disable-print-preview',
      '--disable-setuid-sandbox',
      '--disable-speech-api',
      '--enable-features=SharedArrayBuffer',
      '--hide-scrollbars',
      '--mute-audio',
      '--no-default-browser-check',
      '--no-pings',
      '--no-sandbox',
      '--no-zygote',
      '--disable-extensions',
      '--disable-dev-shm-usage',
      '--no-first-run',
      '--disable-background-networking',
      '--disable-gpu',
      '--disable-software-rasterizer',
    ],
    defaultViewport: {
      deviceScaleFactor: 1,
      hasTouch: false,
      height: 1080,
      isLandscape: true,
      isMobile: false,
      width: 1920,
    },
    ignoreHTTPSErrors: true,
    executablePath:
      process.env.USE_FIREFOX == 'true'
        ? process.env.FIREFOX_PATH
        : process.env.CHROMIUM_PATH,
    // run in shell mode if headless
    headless: true,
    browser: process.env['USE_FIREFOX'] == 'true' ? 'firefox' : 'chrome',
    product: process.env['USE_FIREFOX'] == 'true' ? 'firefox' : 'chrome',
    timeout: 30000,
    dumpio: true,

    // filter out targets
    targetFilter: (target: Target) =>
      target.type() !== 'other' || !!target.url(),
  })) as unknown as Browser

  const version = await browserInstance.version()
  console.log('Browser started', version)

  browserInstance.on('disconnected', () => {
    void handleDisconnection()
  })

  return browserInstance
}

export const closeBrowser = async (): Promise<void> => {
  if (browserInstance) {
    console.log('Closing browser...')
    try {
      await browserInstance.close()
      console.log('Browser closed successfully')
    } catch (error) {
      console.error('Error closing browser:', error)
    } finally {
      browserInstance = null
    }
  } else {
    console.log('No browser instance to close')
  }
}

File: packages/content-fetch-go/internal/browser/browser.go
// Package browser manages a shared headless Chromium instance via chromedp.
package browser

import (
	"context"
	"fmt"
	"log"
	"sync"

	"github.com/chromedp/chromedp"
	"github.com/omnivore-app/omnivore/content-fetch-go/internal/config"
)

// Browser wraps a persistent chromedp browser allocator.
type Browser struct {
	config         *config.Config
	allocCtx    context.Context
	allocCancel context.CancelFunc
	mu          sync.Mutex
}

func New(config *config.Config) *Browser {
	return &Browser{config: config}
}

// allocatorOpts returns the chromedp ExecAllocator options matching the original Puppeteer args.
func (b *Browser) allocatorOpts() []chromedp.ExecAllocatorOption {
	opts := append(chromedp.DefaultExecAllocatorOptions[:],
		chromedp.Flag("autoplay-policy", "user-gesture-required"),
		chromedp.Flag("disable-component-update", true),
		chromedp.Flag("disable-domain-reliability", true),
		chromedp.Flag("disable-print-preview", true),
		chromedp.Flag("disable-setuid-sandbox", true),
		chromedp.Flag("disable-speech-api", true),
		chromedp.Flag("enable-features", "SharedArrayBuffer"),
		chromedp.Flag("hide-scrollbars", true),
		chromedp.Flag("mute-audio", true),
		chromedp.Flag("no-default-browser-check", true),
		chromedp.Flag("no-pings", true),
		chromedp.Flag("no-sandbox", true),
		chromedp.Flag("no-zygote", true),
		chromedp.Flag("disable-extensions", true),
		chromedp.Flag("disable-dev-shm-usage", true),
		chromedp.Flag("no-first-run", true),
		chromedp.Flag("disable-background-networking", true),
		chromedp.Flag("disable-gpu", true),
		chromedp.Flag("disable-software-rasterizer", true),
		chromedp.Flag("ignore-certificate-errors", true),
		chromedp.WindowSize(1920, 1080),
		chromedp.Headless,
	)

	if b.config.ChromiumPath != "" && !b.config.UseFirefox {
		opts = append(opts, chromedp.ExecPath(b.config.ChromiumPath))
	}

	return opts
}

// getAllocator returns the shared browser allocator context, creating it if needed.
func (b *Browser) getAllocator() (context.Context, error) {
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.allocCtx != nil {
		// Check if still alive
		select {
		case <-b.allocCtx.Done():
			// Allocator died, recreate
			b.allocCtx = nil
			b.allocCancel = nil
		default:
			return b.allocCtx, nil
		}
	}

	log.Println("Starting chromedp browser allocator")
	ctx, cancel := chromedp.NewExecAllocator(context.Background(), b.allocatorOpts()...)
	b.allocCtx = ctx
	b.allocCancel = cancel
	return ctx, nil
}

// NewContext creates a new browser tab context using the shared allocator.
func (b *Browser) NewContext() (context.Context, context.CancelFunc, error) {
	allocCtx, err := b.getAllocator()
	if err != nil {
		return nil, nil, fmt.Errorf("get allocator: %w", err)
	}

	ctx, cancel := chromedp.NewContext(allocCtx)
	return ctx, cancel, nil
}

// Close shuts down the browser.
func (b *Browser) Close() {
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.allocCancel != nil {
		b.allocCancel()
		b.allocCancel = nil
		b.allocCtx = nil
		log.Println("Browser closed")
	}
}
